### PR TITLE
Add lane departure alert for Toyota

### PR DESCRIPTION
**Pull request by arne182**

### CODE REVIEW

**Comment by mtourangeau** [comment]
Can I test this?

**Comment by arne182** [comment]
Yes @mtourangeau. It is ready now for testing.

**Comment by arne182** [comment]
Requirements. Must drive above 40 kph and cross a lane line without blinking, or at least get to within 15cm of it.

**Comment by arne182** [comment]
Still getting a raise IndexError('Out of bounds') from path_plan.pathPlan.lPoly[3]

File: selfdrive/controls/controlsd.py [hunk 1]
```diff
@@ -307,6 +307,8 @@ def data_send(plan, path_plan, CS, CI, CP, VM, state, events, actuators, v_cruis
     CC.hudControl.leadVisible = plan.hasLead
     CC.hudControl.rightLaneVisible = bool(path_plan.pathPlan.rProb > 0.5)
     CC.hudControl.leftLaneVisible = bool(path_plan.pathPlan.lProb > 0.5)
+    CC.hudControl.rightLaneDepart = bool(path_plan.pathPlan.rPoly[3] > -1.11 and not CS.rightBlinker) and CS.vEgo > 12.5 # Speed needs to be above 12.5m/s for LDA and only if blinker if off
```
Comment (legonigel) [inline selfdrive/controls/controlsd.py]: Where does the magic value `-1.11`  and `1.05` below come from?

Comment (arne182) [reply selfdrive/controls/controlsd.py]: They are test values from center of car to model line

Comment (arne182) [reply selfdrive/controls/controlsd.py]: Taking into account that the car is 2m wide and the detected lines don't seem to go underneath 1m

**Comment by pd0wm** [comment]
I like the feature! 

Since this is only a UI change, and no alerts/sounds are triggered this can be merged without extensive testing for false positives. If we want to add those add some point we should probably follow the same procedure as for FCW: https://medium.com/@comma_ai/bringing-forward-collision-warnings-to-our-open-source-self-driving-car-7545b6e398cd

If my comments are fixed I'll merge it.

File: selfdrive/car/toyota/carcontroller.py [hunk 2]
```diff
@@ -246,7 +247,7 @@ def update(self, sendcan, enabled, CS, frame, actuators,
       send_ui = False
 
     if (frame % 100 == 0 or send_ui) and ECU.CAM in self.fake_ecus:
-      can_sends.append(create_ui_command(self.packer, steer, sound1, sound2, left_line, right_line))
+      can_sends.append(create_ui_command(self.packer, steer, sound1, sound2, left_line, right_line, leftLane_Depart, rightLane_Depart))
```
Comment (pd0wm) [inline selfdrive/car/toyota/carcontroller.py]: Can you follow the python naming convention? So `left_lane_depart`?

**Review by pd0wm — CHANGES_REQUESTED** [review]

File: selfdrive/controls/controlsd.py [hunk 3]
```diff
@@ -307,6 +307,14 @@ def data_send(plan, path_plan, CS, CI, CP, VM, state, events, actuators, v_cruis
     CC.hudControl.leadVisible = plan.hasLead
     CC.hudControl.rightLaneVisible = bool(path_plan.pathPlan.rProb > 0.5)
     CC.hudControl.leftLaneVisible = bool(path_plan.pathPlan.lProb > 0.5)
+    try:
+      CC.hudControl.rightLaneDepart = bool(path_plan.pathPlan.rPoly[3] > -1.11 and not CS.rightBlinker) and CS.vEgo > 12.5 and bool(path_plan.pathPlan.rProb > 0.5) # Speed needs to be above 12.5m/s for LDA and only if blinker if off
```
[the remark this answers is not in the document]
Comment (arne182) [reply selfdrive/controls/controlsd.py]: This is already the case. and bool(path_plan.pathPlan.rProb > 0.5) is at the end of the if.

File: selfdrive/controls/controlsd.py [hunk 4]
```diff
@@ -307,6 +307,14 @@ def data_send(plan, path_plan, CS, CI, CP, VM, state, events, actuators, v_cruis
     CC.hudControl.leadVisible = plan.hasLead
     CC.hudControl.rightLaneVisible = bool(path_plan.pathPlan.rProb > 0.5)
     CC.hudControl.leftLaneVisible = bool(path_plan.pathPlan.lProb > 0.5)
+    try:
```
[the remark this answers is not in the document]
Comment (arne182) [reply selfdrive/controls/controlsd.py]: Included all modifications.

File: selfdrive/car/toyota/carcontroller.py (hunk 2 above)
Comment (arne182) [reply selfdrive/car/toyota/carcontroller.py]: right_lane_depart and left_lane_depart to conform with python naming convention has been done both in carcontroller and in toytoacan

File: selfdrive/controls/controlsd.py [hunk 5]
```diff
@@ -307,6 +307,10 @@ def data_send(plan, path_plan, CS, CI, CP, VM, state, events, actuators, v_cruis
     CC.hudControl.leadVisible = plan.hasLead
     CC.hudControl.rightLaneVisible = bool(path_plan.pathPlan.rProb > 0.5)
     CC.hudControl.leftLaneVisible = bool(path_plan.pathPlan.lProb > 0.5)
+    if len(list(path_plan.pathPlan.rPoly)) == 4:
+      CC.hudControl.rightLaneDepart = bool(path_plan.pathPlan.rPoly[3] > -1.11 and not CS.rightBlinker) and CS.vEgo > 12.5 and bool(path_plan.pathPlan.rProb > 0.5) # Speed needs to be above 12.5m/s for LDA and only if blinker if off
```
Comment (pd0wm) [inline selfdrive/controls/controlsd.py]: Can you wrap the whole expression into one bool?

**Comment by pd0wm** [comment]
Thanks for fixing my comments. I added one more small comment.

In the meantime, I'll figure out while CI breaks.

File: selfdrive/controls/controlsd.py (hunk 5 above)
Comment (arne182) [reply selfdrive/controls/controlsd.py]: Done.

Comment (arne182) [reply selfdrive/controls/controlsd.py]: Do you mean add the if statement check into the bool as well?

**Pull request merged**